src/Bubble.vala

```diff
@@ -46,32 +54,26 @@ public class Notifications.Bubble : AbstractBubble {
                 break;
         }
 
-        bool default_action = false;
-        bool has_actions = notification.actions.length > 0;
-
-        for (int i = 0; i < notification.actions.length; i += 2) {
-            if (notification.actions[i] == "default") {
-                default_action = true;
-                break;
-            }
-        }
-
-        contents.action_invoked.connect ((action_key) => {
-            action_invoked (action_key);
-            dismiss ();
-        });
-
         button_release_event.connect ((event) => {
-            if (default_action) {
-                action_invoked ("default");
-                dismiss ();
-            } else if (notification.app_info != null && !has_actions) {
-                try {
-                    notification.app_info.launch (null, null);
+            if (action_name != null) {
```

Comment:
Just an idea: Maybe it would be easier to do something like:
```vala
unowned var group = get_action_group (id);
if (group != null && group.has_action ("default") {
    group.activate_action ("default", see other comment);
} else { 
...
``` 
So there would be no need to implement `Gtk.Actionable` or have a `default_action_name`, etc.
Or did I miss something/is this futureproofing?

Reply:
futureproof, the idea is to have the Gtk and Portal interfaces also directly implemented in the daemon.

Gtk.Actionable/GLib.ActionGroup abstract the diferrences between they, so we can isolate the servers from the GUI.